### PR TITLE
Skip drain when node unready

### PR DESCRIPTION
**Pull request by vikaschoudhary16**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note

```

### CODE REVIEW

**Comment by enxebre** [comment]
When deleting a previously stopped machine deletion hangs for ever on draining

- How to reproduce
1- provoke involuntary disruption to an instance, e.g stop it.
2- try to drain the now unhealthy and unreachable node. Evicting API always returns success. But the pods do not go away.
3- We hang forever waiting for deletion https://github.com/openshift/kubernetes-drain/blob/master/drain.go#L446 (edited)
Deleting a previously stopped machine hangs forever.

- Why?
This can happen for two reasons:
-deadlock PDB
-stateful pod https://kubernetes.io/docs/tasks/run-application/force-delete-stateful-set-pod /

we are calling eviction API which succeeds (because the PDB is granted because there isn't any and the pod is scheduled for the deletion - it actually has a deletionTimestamp) but then waitForDeletion never succeeds because the pod is stateful (has local storage) and those are not allowed to be removed from the API server in the case of an unreachable node

File: pkg/controller/machine/controller.go [hunk 1]
```diff
@@ -197,7 +197,7 @@ func (r *ReconcileMachine) Reconcile(request reconcile.Request) (reconcile.Resul
 		// can be unlinked from a machine when the node goes NotReady and is removed
 		// by cloud controller manager. In that case some machines would never get
 		// deleted without a manual intervention.
-		if _, exists := m.ObjectMeta.Annotations[ExcludeNodeDrainingAnnotation]; !exists && m.Status.NodeRef != nil {
+		if _, exists := m.ObjectMeta.Annotations[ExcludeNodeDrainingAnnotation]; !exists && m.Status.NodeRef != nil && r.isNodeReady(ctx, m.Status.NodeRef.Name) {
```
Comment (enxebre) [inline pkg/controller/machine/controller.go:200]: if anything I think we should probably check the node is unreachable, still not sure we want to hard delete

**Comment by enxebre** [comment]
Some proposal could be:
Option 1:
-machine controller never hard delete if there's no annotation, drains hangs if it's not able to succeed, requires manual intervention.
-As a user you have MHC, which has some opinions and makes automatic decisions for you and will skip drain it required.

Option 2:
We should account for this use case at the machine controller level and consider to skip drain and let deletion move forward when the node is unreachable? - This is a sensitive scenario as the stateful pod could be user application critical data

/hold

**Comment by bison** [comment]
I think drain needs to be configurable at the machine-controller level. At a minimum we should respect some kind of `drain-timeout` annotation. If a drain hasn't completed within that time, proceed with the deletion. This matches the behavior of the `kubectl drain` command.

The question then becomes what the default value should be. I think it makes sense to match `kubectl drain` again and default to zero, which would mean no timeout. Possibly MachineHealthCheck could set a custom timeout.

**Review by michaelgugino — CHANGES_REQUESTED** [review]
Never force delete a machine, never skip draining.  We have a lot of assumptions built in multiple places around draining behavior.

If draining is broke, admin needs to intervene.  This would be a very edge-case, and the exact thing we should be alarming for with metrics 'machine with delete timestamp really old' kind of thing.

**Comment by vikaschoudhary16** [comment]
thanks a lot @bison and @michaelgugino for sharing thoughts.
If i am not mistaken, Brad suggesting along the lines of the option1. 

@michaelgugino agree to what you said regarding manual intervention. I see option1 as "your suggestion" + "configuration option at MHC, if enabled, would remediate by skipping drain in case of situation where drain is impossible to succeed like kubelet down"

**Comment by eparis** [comment]
Do we want to pursue this in 4.2? Did we do it in 4.3? Should we just close this and look toward future releases?

**Comment by enxebre** [comment]
> Do we want to pursue this in 4.2? Did we do it in 4.3? Should we just close this and look toward future releases?

Yes, please let's re-open against master if relevant. fwiw a more generic approach is being proposed here https://groups.google.com/forum/#!topic/kubernetes-sig-cli/f4lLTdg0LsE

**Pull request closed without merging**